quic/s2n-quic-core/src/query.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides `Query` and `QueryMut` traits, which are used for querying
+//! and executing functions on several different providers. This includes
+//! [`Subscriber::ConnectionContext`](crate::event::Subscriber::ConnectionContext)
+//! on a Subscriber and the [`Sender`](crate::datagram::Sender) and
+//! [`Receiver`](crate::datagram::Receiver) types on a datagram [`Endpoint`](crate::datagram::Endpoint).
+
+use core::marker::PhantomData;
+
+pub trait Query {
+    fn execute(&mut self, context: &dyn core::any::Any) -> ControlFlow;
+}
+
+pub trait QueryMut {
+    fn execute_mut(&mut self, context: &mut dyn core::any::Any) -> ControlFlow;
+}
+
+/// Used to tell a query whether it should exit early or go on as usual.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ControlFlow {
```

Comment:
While you're here... We should be able to use the official ControlFlow enums now that we've bumped our MSRV

https://doc.rust-lang.org/stable/std/ops/enum.ControlFlow.html. this can just be a type alias instead:

```suggestion
pub type ControlFlow<B = (), C = ()> = core::ops::ControlFlow<B, C>;
```

Reply:
talked offline; going to look into this later